internal/command/meta_backend.go

```diff
@@ -990,6 +991,31 @@ func (m *Meta) backend_C_r_s(c *configs.Backend, cHash int, sMgr *clistate.Local
 					diags = diags.Append(fmt.Errorf(errBackendMigrateLocalDelete, err))
 					return nil, diags
 				}
+				// When migrating from local to TFC we do not need the original state file to remain
+				// on disk. This will look to make sure that it has been backed up and emptied, and
+				// then deletes it.
+				if c.Type == "cloud" {
+					if localS, ok := localState.(*statemgr.Filesystem); ok {
+						originalFile, err := os.Stat(localS.Path())
+						if err != nil {
+							diags = diags.Append(fmt.Errorf(errBackendMigrateLocalDelete, err))
+							return nil, diags
+						}
+						backupFile, err := os.Stat(localS.BackupPath())
+						if err != nil {
+							diags = diags.Append(fmt.Errorf(errBackendMigrateLocalDelete, err))
+							return nil, diags
+						}
+						if originalFile.Size() == 0 && backupFile.Size() > 0 {
+							log.Printf("[TRACE] Meta.Backend: deleting %s", localS.Path())
+							if err := os.Remove(localS.Path()); err != nil {
+								diags = diags.Append(fmt.Errorf(errBackendMigrateLocalDelete, err))
+								return nil, diags
+							}
+						}
```

Comment:
If you haven't already I would suggest testing how this behaves on Windows.

The Windows filesystem abstraction has some significantly different ideas about the interactions between different simultaneous operations on a file than Unix systems do, and in particular I'm concerned that we might get here while we're holding a state lock, which on Windows is implemented as [an exclusive lock on the state file itself](https://github.com/hashicorp/terraform/blob/70db495d001d8ba410b0fad190554b461ae42263/internal/states/statemgr/filesystem_lock_windows.go#L37-L44), which might prevent removing the directory entry even though we _are_ able to mutate the file contents through our already-open handle.

I'm not _certain_ about how this will interact, or indeed about whether we will actually be holding the lock when we get here, but the interactions between locking and other operations on the state file have been a challenge on Windows in the past, and our current approach was shaped by what was possible under the intersection of Unix-style and Windows-style filesystem behaviors.

The way the diff is shaped makes it hard to see, but [the addition of locking is when we switched from deleting the file to just truncating it](https://github.com/hashicorp/terraform/commit/6162cde6ffde8436ff82aa42b50f7fe7d44fb9e1), which is why I'm fearful of re-introducing the deletion at a different layer now. Here's how it was _before_ introducing locking:

https://github.com/hashicorp/terraform/blob/cc0712edabdbd773f3d4168868410924473bdd3a/state/local.go#L45-L53

That original locking commit actually initially broke the behavior by making it ignore `nil` writes altogether:

https://github.com/hashicorp/terraform/blob/6162cde6ffde8436ff82aa42b50f7fe7d44fb9e1/state/local.go#L130-L133

...but then [a later commit in the same PR](https://github.com/hashicorp/terraform/commit/8f7f1917f23f8b0ac34db3f7e3327a5ac52bb6d4) (#11187) landed on the behavior we still have today, of truncating the file. Note that the commit message specifically addresses why we need to truncate instead of delete:

```
Remove state file data when writing a nil state

The old behavior in this situation was to simply delete the file. Since
we now have a lock on this file we don't want to close or delete it, so
instead truncate the file at offset 0.

Fix a number of related tests
```